grace-core/src/Grace/Import.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ViewPatterns #-}
+
+{- | This module contains the functions and types that power to URI-base imports
+-}
+
+module Grace.Import
+    ( Import(..)
+    , ImportCallback
+    , Resolver(..)
+    , resolverToCallback
+    , ImportError(..)
+    ) where
+
+import Control.Exception.Safe (Exception(..), throw)
+import Grace.Location (Location)
+import Grace.Pretty (Pretty(..))
+import Grace.Syntax (Syntax)
+import Text.URI (URI)
+
+{- | A reference to external source code that will be imported by the
+     interpreter.
+-}
+data Import
+    = File FilePath
+    | URI URI
+
+instance Pretty Import where
+    pretty (File file) = pretty file
+    pretty (URI uri) = pretty uri
+
+-- | Type of the callback function used to resolve URI imports
+type ImportCallback = URI -> IO (Syntax Location Import, Maybe FilePath)
```

Comment:
I usually try to avoid type synonyms, preferring either the fully expanded type or a `newtype`.  Usually the way I decide is based on whether or not the user should care about the real underlying type or not:

* If they care, then expand out the type and don't use a synonym

  The reason why is that it will mean less indirection for the person reading the type and trying to figure out how to use it

* If they don't care (i.e. they treat it as abstract), then use a `newtype`

  The reason why is because then it will improve inferred types and error messages

Reply:
Ok, let's go with a `newtype` then.